url1a41b8e9f2_mr1azl_INF345_tgz-pJ8-TestParserJ8/src/xx/MonCompiler.java

```diff
@@ -0,0 +1,2234 @@
+	 * @param right Object
+	 * @return Object
+	 * @see xx.Compiler#notEqual(Object, Object)
+	 */
+	@Override
+	public Object notEqual(Object left, Object right) {
+		//System.out.println(getCurrentMethodName() + left + "\t" + right); // TODO
+		NodeTest ret = new NodeTest();
+		boolean val = false;
+		ret.setNodeType(NODE_TYPE_BOOLEAN);
+		if (((NodeTest) left).getNodeType() == NODE_TYPE_RELATIVE_LOCATIONPATH
+				|| ((NodeTest) right).getNodeType() == NODE_TYPE_RELATIVE_LOCATIONPATH) {
+
+			ret.getFunctionCalls().put("notEqual",
+					new NodeTest[] { ((NodeTest) left), ((NodeTest) right) });
+			ret.setNodeType(NODE_TYPE_RELATIVE_LOCATIONPATH);
+		} else {
+			String s1 = ((NodeTest) left).getNodeValue();
+			String s2 = ((NodeTest) right).getNodeValue();
+			val = !s1.equals(s2);
+		}
+
+		ret.setNodeValue(Utils.asString(val));
+		return ret;
+	}
+
+	/**
+	 * TODO
+	 * @param argument Object
+	 * @return Object
+	 * @see xx.Compiler#minus(Object)
+	 */
+	@Override
+	public Object minus(Object argument) {
+		//System.out.println(getCurrentMethodName());
+		return "minus";
+	}
+
+	/**
+	 * TODO
+	 * @param qName Object
+	 * @return Object
+	 * @see xx.Compiler#variableReference(Object)
+	 */
+	@Override
+	public Object variableReference(Object qName) {
+		//System.out.println(getCurrentMethodName()); // TODO
+		return "variablereferenec";
+	}
+
+	/**
+	 * Ajout l'appel a la fonction dans la stack
+	 * @param code int
+	 * @param args Object[]
+	 * @return Object
+	 * @see xx.Compiler#function(int, Object[])
+	 */
+	@Override
+	public Object function(int code, Object[] args) {
+		NodeTest ret = new NodeTest();
+		boolean val = false;
+
+		ret.getFunctionCalls().put("" + code, toNodeTestArray(args));
+		ret.setNodeType(NODE_TYPE_RELATIVE_LOCATIONPATH);
+
+		ret.setNodeValue(Utils.asString(val));
+		return ret;
+	}
+
+	/**
+	 * TODO
+	 * @param name Object
+	 * @param args Object[]
+	 * @return Object
+	 * @see xx.Compiler#function(Object, Object[])
+	 */
+	@Override
+	public Object function(Object name, Object[] args) {
+		//System.out.println(getCurrentMethodName()); // TODO
+		//System.out.println(getCurrentMethodName() + "code " + name); // TODO
+		return "fonction2";
+	}
+
+	/**
+	 * Si aucune argument n'est un chemin relative retourn le et logic des
+	 * parametre sinon ajoute un appel a la fonction dans la stack
+	 * @param arguments Object[]
+	 * @return Object
+	 * @see xx.Compiler#and(Object[])
+	 */
+	@Override
+	public Object and(Object[] arguments) {
+		NodeTest ret = new NodeTest();
+		NodeTest[] nodes = toNodeTestArray(arguments);
+		ret.setNodeType(NODE_TYPE_BOOLEAN);
+		if (hasType(nodes, NODE_TYPE_RELATIVE_LOCATIONPATH)) {
+
+			ret.getFunctionCalls().put("and", nodes);
+			ret.setNodeType(NODE_TYPE_RELATIVE_LOCATIONPATH);
+		} else {
+			boolean val = true;
+			for (int i = 0; i < nodes.length; i++) {
+				val = val && Utils.asBoolean(nodes[i].getNodeValue());
+			}
+			ret.setNodeValue(Utils.asString(val));
+		}
+		return ret;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 515:529, 580:595.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.